src/twisted/web/http_headers.py

```diff
@@ -112,16 +108,39 @@ def __cmp__(self, other):
 
     def _encodeName(self, name: Union[str, bytes]) -> bytes:
         """
-        Encode the name of a header (eg 'Content-Type') to an ISO-8859-1 encoded
-        bytestring if required.
+        Encode the name of a header (eg 'Content-Type') to an ISO-8859-1
+        encoded bytestring if required.  It will be canonicalized and
+        whitespace-sanitized.
 
         @param name: A HTTP header name
 
         @return: C{name}, encoded if required, lowercased
         """
+        if canonicalName := self._canonicalHeaderCache.get(name, None):
+            return canonicalName
+
         if isinstance(name, str):
-            return name.lower().encode("iso-8859-1")
-        return name.lower()
+            bytes_name = name.encode("iso-8859-1")
+        else:
+            bytes_name = name
+
+        if bytes_name.lower() in self._caseMappings:
+            # Some headers have special capitalization:
+            result = self._caseMappings[bytes_name.lower()]
+        else:
+            result = _sanitizeLinearWhitespace(
+                b"-".join([word.capitalize() for word in bytes_name.split(b"-")])
+            )
+
+        # In general, we should only see a very small number of header
+        # variations in the real world, so caching them is fine. However, an
+        # attacker could generate infinite header variations to fill up RAM, so
+        # we cap how many we cache. The performance degradation from lack of
+        # caching won't be that bad, and legit traffic won't hit it.
+        if len(self._canonicalHeaderCache) < 10_000:
```

Comment:
can you make this a parameter, mostly so that we can get the test coverage for the negative branch?  (But also, threading it through so that users can actually customize the value from the client or server would be nice, we tend to be pretty bad about presenting tunables properly.)

Reply:
I am not sure it needs to be tunable? There's only so many HTTP headers in the real world. So just need it to be large enough that any real-world usage never runs out, and small enough that memory usage isn't a big deal. Could make it 100K by default just as as easily.

I'll start with a private class attribute and if anyone ever complains it can be made public.